File: programs/data_store/src/instructions/roles.rs
use anchor_lang::prelude::*;

use crate::{states::Store, utils::internal};

#[derive(Accounts)]
pub struct CheckRole<'info> {
    pub authority: Signer<'info>,
    pub store: AccountLoader<'info, Store>,
}

/// Verify that the `authority` has the given role in the given `store`.
pub fn check_role(ctx: Context<CheckRole>, role: String) -> Result<bool> {
    ctx.accounts
        .store
        .load()?
        .has_role(ctx.accounts.authority.key, &role)
}

/// Verify that the `user` is an admin of the given `store`.
pub fn check_admin(ctx: Context<CheckRole>) -> Result<bool> {
    Ok(ctx
        .accounts
        .store
        .load()?
        .is_authority(ctx.accounts.authority.key))
}

#[derive(Accounts)]
pub struct HasRole<'info> {
    pub store: AccountLoader<'info, Store>,
}

/// Verify that the `authority` has the given role in the given `store` without signing.
pub fn has_role(ctx: Context<HasRole>, authority: Pubkey, role: String) -> Result<bool> {
    ctx.accounts.store.load()?.has_role(&authority, &role)
}

/// Verify that the `user` is an admin of the given `store` without signing.
pub fn has_admin(ctx: Context<HasRole>, authority: Pubkey) -> Result<bool> {
    Ok(ctx.accounts.store.load()?.is_authority(&authority))
}

#[derive(Accounts)]
pub struct EnableRole<'info> {
    pub authority: Signer<'info>,
    #[account(mut)]
    pub store: AccountLoader<'info, Store>,
}

/// Enable the given role in the data store.
pub fn enable_role(ctx: Context<EnableRole>, role: String) -> Result<()> {
    ctx.accounts.store.load_mut()?.enable_role(&role)
}

impl<'info> internal::Authentication<'info> for EnableRole<'info> {
    fn authority(&self) -> &Signer<'info> {
        &self.authority
    }

    fn store(&self) -> &AccountLoader<'info, Store> {
        &self.store
    }
}

#[derive(Accounts)]
pub struct DisableRole<'info> {
    pub authority: Signer<'info>,
    #[account(mut)]
    pub store: AccountLoader<'info, Store>,
}

/// Disable the given role in the data store.
pub fn disable_role(ctx: Context<DisableRole>, role: String) -> Result<()> {
    ctx.accounts.store.load_mut()?.disable_role(&role)
}

impl<'info> internal::Authentication<'info> for DisableRole<'info> {
    fn authority(&self) -> &Signer<'info> {
        &self.authority
    }

    fn store(&self) -> &AccountLoader<'info, Store> {
        &self.store
    }
}

#[derive(Accounts)]
pub struct GrantRole<'info> {
    pub authority: Signer<'info>,
    #[account(mut)]
    pub store: AccountLoader<'info, Store>,
}

/// Grant a role to the user.
pub fn grant_role(ctx: Context<GrantRole>, user: Pubkey, role: String) -> Result<()> {
    ctx.accounts.store.load_mut()?.grant(&user, &role)
}

impl<'info> internal::Authentication<'info> for GrantRole<'info> {
    fn authority(&self) -> &Signer<'info> {
        &self.authority
    }

    fn store(&self) -> &AccountLoader<'info, Store> {
        &self.store
    }
}

#[derive(Accounts)]
pub struct RevokeRole<'info> {
    pub authority: Signer<'info>,
    #[account(mut)]
    pub store: AccountLoader<'info, Store>,
}

/// Revoke a role to the user.
pub fn revoke_role(ctx: Context<RevokeRole>, user: Pubkey, role: String) -> Result<()> {
    ctx.accounts.store.load_mut()?.revoke(&user, &role)
}

impl<'info> internal::Authentication<'info> for RevokeRole<'info> {
    fn authority(&self) -> &Signer<'info> {
        &self.authority
    }

    fn store(&self) -> &AccountLoader<'info, Store> {
        &self.store
    }
}

File: ts/tests/store/roles.ts
import { AnchorError } from "@coral-xyz/anchor";
import { createRolesPDA } from "../../utils/data";
import { expect, getAddresses, getPrograms, getProvider, getUsers } from "../../utils/fixtures";
import { PublicKey } from "@solana/web3.js";

describe("store: Roles", () => {
    const provider = getProvider();
    const { signer0 } = getUsers();
    const { dataStore } = getPrograms();

    const otherRole = "OTHER";

    let dataStoreAddress: PublicKey;
    let signer0Roles: PublicKey;
    let providerRoles: PublicKey;
    before(async () => {
        ({ dataStoreAddress } = await getAddresses());
        [signer0Roles] = createRolesPDA(dataStoreAddress, signer0.publicKey);
        [providerRoles] = createRolesPDA(dataStoreAddress, provider.publicKey);

        await dataStore.methods.enableRole(otherRole).accounts({
            authority: provider.publicKey,
            store: dataStoreAddress,
        }).rpc();
    });

    after(async () => {
        await dataStore.methods.disableRole(otherRole).accounts({
            authority: provider.publicKey,
            store: dataStoreAddress,
        }).rpc();
    });


    it("check admin success", async () => {
        const isAdmin: boolean = await dataStore.methods.hasAdmin(provider.publicKey).accounts({
            store: dataStoreAddress,
        }).view();
        expect(isAdmin).true;
    });

    it("check admin failure", async () => {
        const isAdmin = await dataStore.methods.hasAdmin(signer0.publicKey).accounts({
            store: dataStoreAddress,
        }).view();
        expect(isAdmin).false;
    });

    it("should fail to transfer store authority without admin role", async () => {
        await expect(dataStore.methods.transferStoreAuthority(signer0.publicKey).accounts({
            authority: signer0.publicKey,
            store: dataStoreAddress,
        }).signers([signer0]).rpc()).rejectedWith(AnchorError, "Not an admin");
    });

    it("transfer back and forth store authority", async () => {
        await dataStore.methods.transferStoreAuthority(signer0.publicKey).accounts({
            authority: provider.publicKey,
            store: dataStoreAddress,
        }).rpc();
        await dataStore.methods.transferStoreAuthority(provider.publicKey).accounts({
            authority: signer0.publicKey,
            store: dataStoreAddress,
        }).signers([signer0]).rpc();
        const authority = (await dataStore.account.store.fetch(dataStoreAddress)).authority;
        expect(authority.equals(provider.publicKey));
    });

    it("should fail to enable role without admin role", async () => {
        await expect(dataStore.methods.enableRole("FOO").accounts({
            authority: signer0.publicKey,
            store: dataStoreAddress,
        }).signers([signer0]).rpc()).to.rejectedWith(AnchorError, "Not an admin");
    });

    it("should fail to disable role without admin role", async () => {
        await expect(dataStore.methods.disableRole("CONTROLLER").accounts({
            authority: signer0.publicKey,
            store: dataStoreAddress,
        }).signers([signer0]).rpc()).to.rejectedWith(AnchorError, "Not an admin");
    });

    it("enable and disable a new role", async () => {
        await expect(dataStore.methods.enableRole("FOO").accounts({
            authority: provider.publicKey,
            store: dataStoreAddress,
        }).rpc());
        await expect(dataStore.methods.disableRole("FOO").accounts({
            authority: provider.publicKey,
            store: dataStoreAddress,
        }).rpc());
    });

    it("should fail to enable or disable role without correct roles", async () => {
        await expect(dataStore.methods.enableRole("FOO").accounts({
            authority: signer0.publicKey,
            store: dataStoreAddress,
        }).signers([signer0]).rpc()).to.rejectedWith(AnchorError, "Not an admin");

        await expect(dataStore.methods.enableRole("FOO").accounts({
            authority: signer0.publicKey,
            store: dataStoreAddress,
        }).signers([signer0]).rpc()).to.rejectedWith(AnchorError, "Not an admin");
    });

    it("grant, check and revoke a role to user", async () => {
        await dataStore.methods.grantRole(signer0.publicKey, otherRole).accountsPartial({
            authority: provider.publicKey,
            store: dataStoreAddress,
        }).rpc();

        {
            const hasRole = await dataStore.methods.hasRole(signer0.publicKey, otherRole).accounts({
                store: dataStoreAddress,
            }).view();
            expect(hasRole).true;
        }

        await dataStore.methods.revokeRole(signer0.publicKey, otherRole).accountsPartial({
            authority: provider.publicKey,
            store: dataStoreAddress,
        }).rpc();

        {
            const hasRole = await dataStore.methods.hasRole(signer0.publicKey, otherRole).accounts({
                store: dataStoreAddress,
            }).view();
            expect(hasRole).false;
        }

        await expect(dataStore.methods.grantRole(signer0.publicKey, otherRole).accountsPartial({
            authority: signer0.publicKey,
            store: dataStoreAddress,
        }).signers([signer0]).rpc()).to.rejectedWith(AnchorError, "Not an admin");

        await expect(dataStore.methods.revokeRole(signer0.publicKey, otherRole).accountsPartial({
            authority: signer0.publicKey,
            store: dataStoreAddress,
        }).signers([signer0]).rpc()).to.rejectedWith(AnchorError, "Not an admin");
    });
});
